Add a [Print Ltacs] vernacular

cpitclaudel opened the pull request:
I quickly mentioned this in the last Coq GT; there does not seem to be a way to print out the signature of all defined tactics. Such a feature would be very convenient in the context of IDEs.

This patch adds a `Print Ltacs` vernacular, which prints out the name and signature of all tactics found in `the_tactictab` in the `Nametab` module. In a clean environment, it looks like this (wrapped for space):

```bash
$ bin/coqtop -coqlib . <<< "Print Ltacs." 2>/dev/null | tail -n +2 | column -c 120
absurd _0                       eauto                           is_cofix _0
absurd_hyp H                    econstructor                    is_evar _0
admit                           ediscriminate                   is_fix _0
assumption                      eexact _0                       is_ground _0
auto                            einjection                      is_var _0
autounfold_one                  eleft                           lapply _0
autounfoldify _0                elimtype _0                     left
bapply lemma todo               eright                          not_evar _0
case_eq x                       esimplify_eq                    now_show
casetype _0                     esplit                          c
cofix                           etransitivity                   red
compare _0 _1                   exact_no_check _0               reflexivity
compute                         exfalso                         revgoals
congruence                      f_equal                         right
constr_eq _0 _1                 fail                            setoid_etransitivity
constr_eq_nounivs _0 _1         false_hyp H G                   setoid_reflexivity
constructor                     fauto                           setoid_symmetry
contradict H                    find_equiv H                    shelve
contradiction                   firstorder                      shelve_unifiable
convert_concl_no_check _0       fresh                           simpl
cut _0                          gintuition                      simplify_eq
destauto                        give_up                         split
destr_eq H                      has_evar _0                     subst
destruct_all                    hnf                             swap H
t                               idtac                           symmetry
dintuition                      info_eauto                      tauto
discriminate                    injection                       transitivity _0
dtauto                          instantiate                     trivial
eassumption                     intro                           vm_cast_no_check
easy                            intros                          _0
easy'                           intuition
```

This draft implementation is pretty straightforward (I just had to extend `NAMETREE` to add an `elements` function). There are a few problems and limitations:

* When long arguments do not fit on one line, they are pushed to the next line. Is there a way to avoid the automatic line breaking?

* It only prints user-defined tactics (this is consistent with `Print Ltac [blah]`, but this means tactic notations are not listed). Tactics like `instantiate` or `setoid_symmetry` appear, but they seem to just be user-level aliases of plugins:

  ```bash
  $ bin/coqtop -coqlib . <<< "Print Ltac instantiate."
  Coq < Ltac instantiate := instantiate
  ```

* It's not clear that `Print Ltacs` is the right name for this (it was inspired by `Print Coercions`). The feature could leave in the `Search` namespace, but there's not much more to search for than tactic names, so I'm not sure how useful it would be.

* It's not clear (to me) that Nametab is the right way to look for these things; maybe it is not?

Regardless, the prototype works well, and it could open opportunities for really neat in-IDE completion! (if the `Search Output Name Only` patch gets merged, tactics completion will be only type of completion still missing from `company-coq`).

herbelin commented:
Hi Clément,

Tactic are indeed either defined as "Ltac" definitions or using a notation (primitively in g_tactic.ml4, coq-side using Tactic Notation), with tactics in plugins being either turned into Ltac definitions or primitive notations or both, depending on their exact syntax (to be confirmed by Pierre-Marie).

If you want all of them, either hack the output of both printing the tactic nametab and of "Print Grammar tactic", or add an extra table for the purpose of collecting all names (with manual input for tactics from g_tactic.ml4).

Btw, do you want also the grammar or only the heading name of the tactic? What do you want to do with tactic notation not starting with an ident?

About your questions:
> When long arguments do not fit on one line, they are pushed to the next line. Is there a way to avoid the automatic line breaking?

Using "Set Printing Width" with a very high limit should avoid automatic line breaking.

> It's not clear that Print Ltacs is the right name for this

As far as it is the nametab that you are requesting, you get tactics whose syntax follows the regular syntax of defined Ltac tactics. So "Ltac" is ok with this respect. But plural is strange. So "Print Ltac Entries", or "Print Ltac Definitions", would fit better, but, this is anyway abstract discussion since you probably want something different from the list of defined Ltac tactics.

Hugo

cpitclaudel commented:
Thanks for your comments Hugo! 

There are essentially two reasons to list tactics and tactic notations: completion is one; the other is easier access to documentation. 

* Completion of user-defined tactics works well with this patch. Completion of (documented) basic tactics works well with my scripts to extract patterns and documentation from the manual. However, completion of tactic notations is still lacking. The pointer to `Print Grammar tactic` is very useful; thanks! (I remember Pierre mentioning it at some point, but I forgot about it). In all cases, I prefer to have the whole signature than just the head name; it allows for a much nicer experience (snippets can be inserted with appropriate placeholders, and in the long run one could even imagine static checking of ltac scripts to show warnings when arguments are missing). Tactic notations that do not start with an indent are not a problem, since completion is done in a fuzzy way.
 
* Documentation is trickier. One nice side effect of `Locate Ltac` (and `Locate` in 8.5) is that they allow me to show the source that surrounds the definition of a tactic; that's often where the documentation for that tactic is found (that's also why I emailed the list regarding installing .v files along with .vo):

  ![source-view-ltac](https://cloud.githubusercontent.com/assets/8181630/7537572/e1caf60a-f567-11e4-8bf7-e2aa56bd8def.png)

  Unfortunately, `Locate` does not seem to work for tactic notations (AFAICS), so the documentation / source view part is a bit out of reach for now; still, completion would be very nice

If I understand the output of `Print Grammar tactic` correctly, I should probably just parse the `simple_tactic` section. The format does not seem too complex, either; I guess I can merge the output of this with the list of known tactics from the manual, keeping only the stuff not already found in the manual. In that case, `Print Ltac Definitions` (or even `Print Ltac Signatures`?) may actually be sufficient!

I'll probably take a stab at this this week-end. I would welcome suggestion on alternative approaches though! For example, would it be simpler (or more desirable?) to broaden the meaning of `Print Grammar tactic` to include user-defined tactics as well?

herbelin commented:
> Unfortunately, Locate does not seem to work for tactic notations (AFAICS), so the documentation / source view part is a bit out of reach for now; still, completion would be very nice

If you are courageous, you could try to adapt Locate "piece-of-a-notation" (which works in interp/notation.ml on the table of constr notations) to some Locate Tactic "piece-of-a-tactic-notation" (within the prnotation_tab table in printing/pptactic.ml).

> If I understand the output of Print Grammar tactic correctly, I should probably just parse the simple_tactic section. 

Tactic notations have 6 levels : level 0 is for simple_tactic, other levels are basically for tacticals. So, it depends whether you are interested in tacticals or not.

Very nice work that you are doing. I'll be envious that coqide has it too!

Incidentally, I'm going to push a patch so that Print Ltac tactic uses more user-friendly variables than _0, _1, ...

> For example, would it be simpler (or more desirable?) to broaden the meaning of Print Grammar tactic to include user-defined tactics as well?

I would say no. Print Grammar tactic is really about grammar only. But we could imagine a Print Tactic Entries, or simply Print Tactics which would mix the results of your Print Ltacs and of the lhs of Print Grammar tactic (one problem being that the abstract tree of Print Grammar tactic is not under our control, it is pure text produced by camlp5/4).

cpitclaudel commented:
> If you are courageous, you could try to adapt Locate "piece-of-a-notation" (which works in interp/notation.ml on the table of constr notations) to some Locate Tactic "piece-of-a-tactic-notation" (within the prnotation_tab table in printing/pptactic.ml).

Hmm. That sounds a bit scary :)

>> If I understand the output of Print Grammar tactic correctly, I should probably just parse the simple_tactic section.

> Tactic notations have 6 levels : level 0 is for simple_tactic, other levels are basically for tacticals. So, it depends whether you are interested in tacticals or not.

I think the focus is mostly on simplicity for now, and tacticals are well described in the manual (and short, and easy to type; and they can even be special-cased to add snippets and related goodies).

> Very nice work that you are doing. I'll be envious that coqide has it too!

Thanks!

> Incidentally, I'm going to push a patch so that Print Ltac tactic uses more user-friendly variables than _0, _1, ...

That would be very nice.

>>  For example, would it be simpler (or more desirable?) to broaden the meaning of Print Grammar tactic to include user-defined tactics as well?

> I would say no. Print Grammar tactic is really about grammar only. But we could imagine a Print Tactic Entries, or simply Print Tactics which would mix the results of your Print Ltacs and of the lhs of Print Grammar tactic (one problem being that the abstract tree of Print Grammar tactic is not under our control, it is pure text produced by camlp5/4).

Ok. If we're going to have to parse the textual output of `Print Grammar tactic`, then I think it's best to leave that parsing to client apps and IDEs. Properly parsing `OPT`s may be tricky, but simple cases should be easy to support to begin with. 

I had a quick look at the format, and one thing that may be much easier from the Coq side is to do some unfolding. For example, the line `  | IDENT "destruct"; induction_clause_list` does not give much information as to the actual way in which the induction command is invoked, and `induction_clause_list` is not one of the documented argument types in the `Tactic Notation` documentation. 

Similarly, a printing option that expands notations (except cyclic ones) could be useful (instead of `"a" [ "b" | "c" ]` it would print `"a" "b" | "a" "c"`). If we went that route, then maybe a full-fledged `Print Tactics` that prints everything entirely expanded would make sense (and it would make implementation of IDEs easier). But maybe this is really an IDE problem, and maybe in some case the factored version is more useful.

cpitclaudel commented:
I hacked together a basic parser for the output of `Print Grammar Tactic`; the list of tactics that I obtained is [here](http://pastebin.com/31SFztBU). A few more thoughts about this:

* The main problem with this approach is that it's not clear how far to expand things, as the following examples should make evident:

  ```
  firstorder @{tactic:tactic} @{tactic:firstorder_using} with @{prim:preident+}
  apply @{constr_with_bindings_arg,+} @{in_hyp_as}
  ```

  In both cases one would want to expand `@{tactic:firstorder_using}` and `@{in_hyp_as}`, but not `@{tactic:tactic}`. I'm not sure how much this matters in practice: these tactics are (almost) all properly documented, and the documentation has exactly the right expansion. Hopefully for tactic notations introduced by plugins and users are simpler (A confirmation that most of these patterns can only appear for "core" tactics would be very nice).

Another question: Are there any guarantees on the order in which entries are printed? It would make it easier to spot new entries.

herbelin commented:
> The main problem with this approach is that it's not clear how far to expand things

I have no answer to this. Factorization are usually done empirically for good reasons. But if you see one of them which is abusive or not optimal, you can report about it.

> Hopefully for tactic notations introduced by plugins and users are simpler (A confirmation that most of these patterns can only appear for "core" tactics would be very nice).

There is no reason a priori it is so. If there is maybe less factorization in plugins, it might be because TACTIC EXTEND is not as convenient as camlp5/4's (G)EXTEND for declaring sub-entries, pushing users to instead do some duplications.

> Another question: Are there any guarantees on the order in which entries are printed? It would make it easier to spot new entries.

This is decided by camlp5/4 and I could not find a clear rule about it. BTW, maybe should you check that camlp4 and camlp5 use the same output format for "Print Grammar tactic" and that the format does not change from one version to another.

cpitclaudel commented:
I've added support for tactic notations to company-coq over the week-end:

![tactic-notation-completion](https://cloud.githubusercontent.com/assets/8181630/7566383/77c66ab4-f7c4-11e4-808f-ef1fa309f11a.png)

There are slight differences in the output of `Print Grammar tactic` in 8.4 and 8.5, but nothing too significant. The results look pretty good to me, so I'm not sure if a unified printer for user-defined tactics and tactic notations is really necessary (and I'm not too sure which form it would take). 

Two incremental improvements that I can think of would be:

1. An ability to locate notations (or even better, a way for `Print Grammar tactic` to also print location information, as discussed above).
2. More consistency between notation definitions and the way they are printed back by `Print Grammar tactic`. In the example above, `hyp(blah)` becomes `prim:var`, and `constr(blah)` becomes `constr:constr`. It would be nice in particular to keep argument names.

These two improvements are pretty independent from this patch though.

cpitclaudel commented:
I've updated the patch to rename the command to `Print Ltac Signatures`, and I added documentation. Is this a good name? And is there anything else I should change?

maximedenes commented:
test build

maximedenes commented:
test build

cpitclaudel commented:
Ping? Is there anything I should do to get this merged?

ejgallego commented:
Looking at this PR I have two comments:

- First, indeed I think that there's an issue with discoverability and `Nametab`. Recently I wanted to dump some objects there an indeed I had to write a patch equivalent to Clement's `elements` addition. It looks reasonably that IDES would like to access to some of the content there, but of course I may be ignoring other ways to access this information.

- Second, I'm not sure that a adding a new vernacular everytime we need some information on the coq side may be the best approach from the maintenance point of view. Clément, would a more generic API of the form "query(objects)" + "print(objects)" work for you?

cpitclaudel commented:
Sure, anything would work fine for me, as long as it's accessible from a standard `coqtop`.

ppedrot commented:
I find the Nametab hack quite ugly. You should rather look for the tactics defined in the Tacenv and then filter according to Nametab. I think I'll work a bit further on this patch, e.g. extending it to Tactic Notations before merging. I cannot do it right now, but I'll have a bit of spare time at the end of the week.

cpitclaudel commented:
@ppedrot Thanks for your help! Why do you call it a hack? Regarding tactic notations, can we include a way to tell them apart from actual tactics? This matters because you can already get tactic notations by parsing the output of `Print Grammar tactic`.

mattam82 commented:
@ppedrot How's the hacking going? Should it use val Tacenv.ltac_entries : unit -> ltac_entry KNmap.t
instead?

ppedrot commented:
Reimplemented by commit 45ee3d6.

cpitclaudel commented:
@ppedrot Thanks for getting this done, if only because it was getting a bit ridiculous to have a PR opened for over a year.

Still, I find it disappointing that you decided to trash this code and reimplement everything on your own, without answering my question (what part of the original patch was a hack?) nor @ejgallego's concern that we may be introducing too many new vernacs.

ppedrot commented:
Nametab should not be used to look for defined objects in the environment. It only cares about names, and your use was probably brittle. Furthermore, putting your code in Tacintern was clearly wrong, as the latter is dedicated to tactic internalization and has nothing to do with vernacular commands. As for vernac introduction, I do not know if there is a reasonable way to counter command inflation.

BTW, if you're around Princeton tonight we can discuss it around a beer or similar.

JasonGross commented:
@ppedrot are you at Princeton?  I'd love to join a Coq discussion over food rather than sitting in this dorm room...

cpitclaudel commented:
Sounds good, but my flight lands at 11; see you tomorrow!
